pxr/imaging/hdSt/meshShaderKey.cpp

```diff
@@ -346,28 +350,82 @@ HdSt_MeshShaderKey::HdSt_MeshShaderKey(
     }
 
     bool const ptvsStageEnabled = !PTVS[0].IsEmpty();
+    bool finalTesStageEnabled = false;
 
-    // tessellation control shader
     if (isPrimTypePatches && !ptvsStageEnabled) {
-        TCS[0] = _tokens->instancing;
-        TCS[1] = _tokens->mainPatchCommonTCS;
-        TCS[2] = isPrimTypePatchesBSpline
+
+        // Tessellation control shader.
+        uint8_t tcsIndex = 0;
+        TCS[tcsIndex++] = _tokens->instancing;
+        TCS[tcsIndex++] = _tokens->mainPatchCommonTCS;
+        TCS[tcsIndex++] = isPrimTypePatchesBSpline
                      ? _tokens->mainBSplineQuadTCS
                      : _tokens->mainBoxSplineTriangleTCS;
-        TCS[3] = TfToken();
-
-        // tessellation evaluation shader
-        TES[0] = _tokens->instancing;
-        TES[1] = isPrimTypePatchesBSpline
-                     ? _tokens->mainBezierQuadTES
-                     : _tokens->mainBezierTriangleTES;
-        TES[2] = _tokens->mainVaryingInterpTES;
-        TES[3] = TfToken();
+        TCS[tcsIndex++] = TfToken();
+
+        // Tessellation evaluation shader
+        uint8_t tesIndex = 0;
+        TES[tesIndex++] = _tokens->instancing;
+
+        if (geomStyle == HdMeshGeomStyleEdgeOnSurf) {
```

Comment:
This seems not quite right. Minimally, this should also test `hasBuiltinBarycentrics`, which is the only thing that edges need from GS.

Reply:
Ah, right, when the fragment shader has barycentrics, it can render edges.